sql/analyzer/optimize_joins.go

```diff
@@ -0,0 +1,364 @@
+	return indexExprStr == expr.String()
+}
+
+// Returns the underlying table name for the node given
+func findTableName(node sql.Node) string {
+	var tableName string
+	plan.Inspect(node, func(node sql.Node) bool {
+		switch node := node.(type) {
+		case *plan.ResolvedTable:
+			// TODO: this is over specific, we only need one side of the join to be indexable
+			if it, ok := node.Table.(sql.IndexableTable); ok {
+				tableName = it.Name()
+				return false
+			}
+		}
+		return true
+	})
+
+	return tableName
+}
+
+// Returns the table and field names from the expression given
+func getTableNameFromExpression(expr sql.Expression) (tableName string, fieldName string) {
+	switch expr := expr.(type) {
+	case *expression.GetField:
+		tableName = expr.Table()
+		fieldName = expr.Name()
+	}
+
+	return tableName, fieldName
+}
+
+// index munging
+
+// Assign indexes to the join conditions and returns the sql.Indexes assigned, as well as returning any aliases used by
+// join conditions
+func findJoinIndexes(ctx *sql.Context, a *Analyzer, node sql.Node) ([]sql.Index,  Aliases, error) {
+	a.Log("finding indexes, node of type: %T", node)
+
+	indexSpan, _ := ctx.Span("find_join_indexes")
+	defer indexSpan.Finish()
+
+	var indexes []sql.Index
+	// release all unused indexes
+	defer func() {
+		if indexes == nil {
+			return
+		}
+
+		for _, index := range indexes {
+			a.Catalog.ReleaseIndex(index)
+		}
+	}()
+
+	aliases := make(Aliases)
+	var err error
+
+	fn := func(ex sql.Expression) bool {
+		if alias, ok := ex.(*expression.Alias); ok {
+			if _, ok := aliases[alias.Name()]; !ok {
+				aliases[alias.Name()] = alias.Child
+			}
+		}
+		return true
+	}
+
+	plan.Inspect(node, func(node sql.Node) bool {
+		switch node := node.(type) {
+		case *plan.InnerJoin, *plan.LeftJoin, *plan.RightJoin:
+
+			var cond sql.Expression
+			switch node := node.(type) {
+			case *plan.InnerJoin:
+				cond = node.Cond
+			case *plan.LeftJoin:
+				cond = node.Cond
+			case *plan.RightJoin:
+				cond = node.Cond
+			}
+
+			fn(cond)
+
+			var idxes []sql.Index
+			idxes, err = getJoinIndexes(cond, aliases, a)
+			if err != nil {
+				return false
+			}
+
+			indexes = addManyToSet(indexes, idxes)
+		}
+
+		return true
+	})
+
+	return indexes, aliases, err
+}
+
+func addManyToSet(indexes []sql.Index, toAdd []sql.Index) []sql.Index {
+	for _, i := range toAdd {
+		indexes = addToSet(indexes, i)
+	}
+	return indexes
+}
+
+func addToSet(indexes []sql.Index, index sql.Index) []sql.Index {
+	for _, idx := range indexes {
+		if idx == index {
+			return indexes
+		}
+	}
+
+	return append(indexes, index)
+}
+
+// Returns the left and right indexes for the two sides of the equality expression given.
+func getJoinEqualityIndex(
+		a *Analyzer,
+		e *expression.Equals,
+		aliases map[string]sql.Expression,
+) (leftIdx sql.Index, rightIdx sql.Index) {
+
+	// Only handle column expressions -- evaluable expressions will have already gotten pushed down
```

Comment:
Is this true given the pushdown bailout in `pushdown.go`?

Reply:
Kind of. Refined the comment.